test/IXWebSocketCloseTest.cpp

```diff
@@ -0,0 +1,407 @@
+            bool getCloseRemote();
+
+        private:
+            ix::WebSocket _webSocket;
+            int _port;
+
+            mutable std::mutex _mutexCloseData;
+            uint16_t _closeCode;
+            std::string _closeReason;
+            bool _closeRemote;
+    };
+
+    WebSocketClient::WebSocketClient(int port)
+        : _port(port)
+        , _closeCode(0)
+        , _closeReason(std::string(""))
+        , _closeRemote(false)
+    {
+        ;
+    }
+
+    bool WebSocketClient::isReady() const
+    {
+        return _webSocket.getReadyState() == ix::WebSocket_ReadyState_Open;
+    }
+
+    uint16_t WebSocketClient::getCloseCode()
+    {
+        std::lock_guard<std::mutex> lck(_mutexCloseData);
+        
+        return _closeCode;
+    }
+
+    const std::string& WebSocketClient::getCloseReason()
+    {
+        std::lock_guard<std::mutex> lck(_mutexCloseData);
+
+        return _closeReason;
+    }
+
+    bool WebSocketClient::getCloseRemote()
+    {
+        std::lock_guard<std::mutex> lck(_mutexCloseData);
+        
+        return _closeRemote;
+    }
+
+    void WebSocketClient::stop()
+    {
+        _webSocket.stop();
+    }
+
+    void WebSocketClient::stop(uint16_t code, const std::string& reason)
+    {
+        _webSocket.close(code, reason);
+        _webSocket.stop();
+    }
+
+    void WebSocketClient::start()
+    {
+        std::string url;
+        {
+            std::stringstream ss;
+            ss << "ws://localhost:"
+               << _port
+               << "/";
+
+            url = ss.str();
+        }
+
+        _webSocket.setUrl(url);
+
+        std::stringstream ss;
+        log(std::string("Connecting to url: ") + url);
+
+        _webSocket.setOnMessageCallback(
+            [this](ix::WebSocketMessageType messageType,
+               const std::string& str,
+               size_t wireSize,
+               const ix::WebSocketErrorInfo& error,
+               const ix::WebSocketOpenInfo& openInfo,
+                   const ix::WebSocketCloseInfo& closeInfo)
+            {
+                std::stringstream ss;
+                if (messageType == ix::WebSocket_MessageType_Open)
+                {
+                    log("client connected");
+
+                    _webSocket.disableAutomaticReconnection();
+                }
+                else if (messageType == ix::WebSocket_MessageType_Close)
+                {
+                    log("client disconnected");
+
+                    std::lock_guard<std::mutex> lck(_mutexCloseData);
+                    
+                    _closeCode = closeInfo.code;
+                    _closeReason = std::string(closeInfo.reason);
+                    _closeRemote = closeInfo.remote;
+
+                    _webSocket.disableAutomaticReconnection();
+                }
+                else if (messageType == ix::WebSocket_MessageType_Error)
+                {
+                    ss << "Error ! " << error.reason;
+                    log(ss.str());
+
```

Comment:
I don't like all those repeated disableAutomaticReconnection ... we should only have to specify this once, before we call start. Is that not working ?

Reply:
I'm not certain that disableAutomaticReconnection work too well, it is only used for the server code, I never tested it with the client code that much. We could try to fix it in another PR.

Reply:
No it’s not working because before you won’t start at all,